### PR TITLE
DE50536: ACS: policy evaluation against non-existent acs zone causes 500 response code

### PR DESCRIPTION
**Pull request by VoollaSandeepKumar**
DE50536: ACS: policy evaluation against non-existent acs zone causes 500 response code

### CODE REVIEW

File: service/src/test/java/com/ge/predix/acs/testutils/TestUtils.java [hunk 1]
```diff
@@ -114,4 +115,17 @@ public MockMvcContext createWACWithCustomPOSTRequestBuilder(final WebApplication
         result.setMockMvc(MockMvcBuilders.webAppContextSetup(wac).defaultRequest(result.getBuilder()).build());
         return result;
     }
+
+    public static class TestAssertion {
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/testutils/TestUtils.java:118]: If you want to keep this inner class, I would put a `TODO` saying that we should start moving tests to use this `assertDoesNotThrow` method wherever we use the `try { ... } catch (Exception e) { Assert.fail(); }` pattern.

File: service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java [hunk 2]
```diff
@@ -29,10 +24,15 @@
...
 
+import com.ge.predix.acs.privilege.management.PrivilegeManagementUtility;
+import com.ge.predix.acs.privilege.management.ZoneDoesNotExistException;
 
-@ControllerAdvice(basePackageClasses = { SubjectPrivilegeManagementController.class,
-        ResourcePrivilegeManagementController.class, AttributeConnectorController.class })
+// A ZoneDoesNotExistException exception thrown from any @RestController should be handled by this @ControllerAdvice.
+// Note that /v1/zone* APIs don't throw ZoneDoesNotExistExceptions anywhere in their call hierarchy so we don't need
+// to explicitly exclude them here.
+@ControllerAdvice(annotations = RestController.class)
```
Comment (irinaepshteyn) [inline service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java]: Since this controller advice applies to all controllers, merge it with BaseRestApiControllerAdvice after DE49777 branch is merged.

File: service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java [hunk 3]
```diff
@@ -86,6 +86,15 @@ public void setup() {
     }
 
     @Test(dataProvider = "requestUrlProvider")
+    public void attributeConnectorZoneDoesNotExistException(final String endpointUrl) throws Exception {
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java]: No need to test this with data provider.  One pass is enough.  Also, there is testZoneDoesNotExist() below.  How is it different?

File: service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java [hunk 4]
```diff
@@ -86,6 +86,15 @@ public void setup() {
...
+        createZone1AndAssert();
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java]: No need to setup a zone if we are testing "zone does not exist" use case.

File: service/src/test/java/com/ge/predix/acs/testutils/MockSecurityContext.java [hunk 5]
```diff
@@ -35,7 +35,9 @@ private MockSecurityContext() {
 
     public static void mockSecurityContext(final Zone zone) {
         ZoneOAuth2Authentication acsZoneOAuth = Mockito.mock(ZoneOAuth2Authentication.class);
-        when(acsZoneOAuth.getZoneId()).thenReturn(zone.getSubdomain());
+        if (zone != null) {
+            when(acsZoneOAuth.getZoneId()).thenReturn(zone.getSubdomain());
+        }
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/acs/testutils/MockSecurityContext.java:40]: Should we in addition mock acsZoneOAuth.getZoneId() to return null?

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java [hunk 6]
```diff
@@ -104,12 +104,32 @@ public void setup() {
...
+    @Test
+    public void testPolicyZoneDoesNotExistException() throws Exception {
+        MockSecurityContext.mockSecurityContext(null);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
+
+        MockMvcContext postPolicyEvalContext = this.testUtils
+                .createWACWithCustomPOSTRequestBuilder(this.wac, this.testZone.getSubdomain(), POLICY_EVAL_URL);
+        postPolicyEvalContext.getMockMvc().perform(
+                postPolicyEvalContext.getBuilder().contentType(MediaType.APPLICATION_JSON)
+                        .content("testString")).andExpect(status().isBadRequest());
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: How can we be sure that bad request status was indeed caused by ZoneDoesNotExistException?  We should mock exception throw.

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java [hunk 7]
```diff
@@ -104,12 +104,32 @@ public void setup() {
...
+        MockSecurityContext.mockSecurityContext(this.testZone);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: These mocks are done in beforeMethod().  No need to do it here.

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java [hunk 8]
```diff
@@ -81,13 +83,33 @@ public void setup() throws Exception {
...
+        MockSecurityContext.mockSecurityContext(null);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
+
+        String thisUri = VERSION + "/policy-set/" + this.policySet.getName();
+        // create policy-set in first zone
+        MockMvcContext putContext = this.testUtils.createWACWithCustomPUTRequestBuilder(this.wac,
+                this.testZone.getSubdomain(), thisUri);
+        putContext.getMockMvc().perform(putContext.getBuilder().contentType(MediaType.APPLICATION_JSON)
+                .content(this.objectWriter.writeValueAsString(this.policySet))).andExpect(status()
+                .isUnprocessableEntity());
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: Same concern as above - how can we be sure that Unprocessable Entity  status was indeed caused by ZoneDoesNotExistException? We should mock exception throw.

File: service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java [hunk 9]
```diff
@@ -83,11 +84,31 @@ public void setup() throws Exception {
...
+        MockSecurityContext.mockSecurityContext(null);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
+        BaseResource resource = JSON_UTILS.deserializeFromFile("controller-test/a-resource.json",
+                                                                    BaseResource.class);
+        String thisUri = RESOURCE_BASE_URL + "/%2Fservices%2Fsecured-api";
+        // create resource in first zone
+        MockMvcContext putContext =
+            TEST_UTILS.createWACWithCustomPUTRequestBuilder(this.wac, this.testZone.getSubdomain(), thisUri);
+        putContext.getMockMvc().perform(putContext.getBuilder().contentType(MediaType.APPLICATION_JSON)
+                .content(OBJECT_MAPPER.writeValueAsString(resource))).andExpect(status().isBadRequest());
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java]: Mock ZoneDoesNotExistException throw to make sure this is what caused bad request status.

File: service/src/test/java/com/ge/predix/controller/test/SubjectPrivilegeManagementControllerIT.java [hunk 10]
```diff
@@ -82,16 +85,34 @@
...
+        // NOTE: To throw a ZoneDoesNotExistException, we must ensure that the AcsRequestContext in the
+        //       SpringSecurityZoneResolver class returns a null ZoneEntity
+        MockSecurityContext.mockSecurityContext(null);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
+
+        BaseSubject subject = JSON_UTILS.deserializeFromFile("controller-test/a-subject.json", BaseSubject.class);
+        MockMvcContext putContext = TEST_UTILS.createWACWithCustomPUTRequestBuilder(this.wac,
+                "zoneDoesNotExist", SUBJECT_BASE_URL + '/' + subject.getSubjectIdentifier());
+        putContext.getMockMvc().perform(putContext.getBuilder().contentType(MediaType.APPLICATION_JSON).
+                content(OBJECT_MAPPER.writeValueAsString(subject))).andExpect(status().isBadRequest());
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/SubjectPrivilegeManagementControllerIT.java]: Mock ZoneDoesNotExistException throw to make sure this is what caused bad request status.

File: service/src/test/java/com/ge/predix/acs/testutils/TestUtils.java (hunk 1 above)
Comment (irinaepshteyn) [reply service/src/test/java/com/ge/predix/acs/testutils/TestUtils.java:118]: I don't see it being used anywhere.  Should either use it or remove it.

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
Done initial review.  Please address my comments.

File: service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java (hunk 2 above)
Comment (VoollaSandeepKumar) [reply service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java]: Will fix this once this defect is approved

File: service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java (hunk 3 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java]: As we discussed we are testing once for each controller for Zone does not exist so the data provider remains

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java (hunk 7 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: There is test cleanup after method for which the mock context need to be  set back so that testCleanup runs susccessfully

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java (hunk 6 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: Thats a good question. Now I want to check all the test cases written to make sure bad request indeed caused by ZoneDoesNotExistException. I have changed the test case to assert the message added by the control advice for ZoneDoesNotExistException

File: service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java (hunk 4 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/AttributeConnectorControllerIT.java]: I reverted my changes as there is a test already checking for zoneDoesnotExist with test name TestZoneDoesNotExist

**Review by VoollaSandeepKumar — COMMENTED** [review]
Testing ZoneDoesNotExist Exception by asserting the message of the controller Response

File: service/src/test/java/com/ge/predix/acs/testutils/TestUtils.java (hunk 1 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/acs/testutils/TestUtils.java:118]: I have added a task in triage. in future we can implement this feature for exception handling

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java [hunk 11]
```diff
@@ -94,7 +96,6 @@ public void setup() {
         this.zoneService.upsertZone(this.testZone);
         MockSecurityContext.mockSecurityContext(this.testZone);
         MockAcsRequestContext.mockAcsRequestContext(this.testZone);
-
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: Four lines above (#93-96) could be replaced with one call to TestUtils:
this.testZone = new TestUtils().setupTestZone(...)

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java [hunk 12]
```diff
@@ -104,13 +105,39 @@ public void setup() {
         this.notApplicableAndDenyPolicySets = createNotApplicableAndDenyPolicySets();
     }
 
+    @BeforeMethod
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: This beforeMethod() is redundant.  setup() will create test zone and setup mock context for all tests.  Then in testPolicyZoneDoesNotExistException() you can overwrite security context with null zone at the beginning of the test and restore it to the test zone at the end of the test.

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java [hunk 13]
```diff
@@ -104,13 +105,39 @@ public void setup() {
...
     @AfterMethod
     public void testCleanup() {
         List<PolicySet> policySets = this.policyManagementService.getAllPolicySets();
         policySets.forEach(policySet -> this.policyManagementService.deletePolicySet(policySet.getName()));
     }
 
     @Test
+    public void testPolicyZoneDoesNotExistException() throws Exception {
+        MockSecurityContext.mockSecurityContext(null);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: No need to mock acs request context again.

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java [hunk 14]
```diff
@@ -104,13 +105,39 @@ public void setup() {
...
+                .createWACWithCustomPOSTRequestBuilder(this.wac, this.testZone.getSubdomain(), POLICY_EVAL_URL);
+        ResultActions resultActions = postPolicyEvalContext.getMockMvc().perform(
+                postPolicyEvalContext.getBuilder().contentType(MediaType.APPLICATION_JSON)
+                        .content(OBJECT_MAPPER.writeValueAsString(policyEvalRequest)));
+        resultActions.andReturn().getResponse()
+                .getContentAsString().contentEquals("Zone not found");
+        resultActions.andExpect(status().isBadRequest());
+
+        MockSecurityContext.mockSecurityContext(this.testZone);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: No need to mock acs request context again.

File: service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java (hunk 2 above)
Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java]: DE49777 branch has been approved and merged.  You should be able to merge this controller advice with the BaseRestApiControllerAdvice.

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java [hunk 15]
```diff
@@ -81,13 +84,17 @@ public void setup() throws Exception {
         this.testZone2 = new TestUtils().createTestZone("PolicyMgmtControllerIT2");
         this.zoneService.upsertZone(this.testZone);
         this.zoneService.upsertZone(this.testZone2);
-        MockSecurityContext.mockSecurityContext(this.testZone);
-        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: Lines above (#81, 82, 84, 85) could be replaced with one call to TestUtils:
this.testZone = new TestUtils().setupTestZone(...).

No need to remove security and request contexts setup from here.  Only one test policyZoneDoesNotExistException() could overwrite security context with null zone and restore when done.

Setting up testZone2 could be deferred until the test that is using it -testCreateSamePolicyDifferentZones()

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java [hunk 16]
```diff
@@ -81,13 +84,17 @@ public void setup() throws Exception {
...
         this.policySet = this.jsonUtils.deserializeFromFile("controller-test/complete-sample-policy-set.json",
                 PolicySet.class);
         Assert.assertNotNull(this.policySet, "complete-sample-policy-set.json file not found or invalid");
     }
 
+    @BeforeMethod
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: No need for beforeMethod() as only one test policyZoneDoesNotExistException() could overwrite security context with null zone and restore when done.

File: service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java [hunk 17]
```diff
@@ -83,11 +86,15 @@ public void setup() throws Exception {
         this.testZone2 = TEST_UTILS.createTestZone("ResourceMgmtControllerIT2");
         this.zoneService.upsertZone(this.testZone);
         this.zoneService.upsertZone(this.testZone2);
+    }
+
+    @BeforeMethod
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java]: Refer to comments for PolicyManagementControllerIT regarding setup() and beforeMethod().  Setting up testZone2 could be deferred until the  test that is using it - testSameResourceDifferentZones()

File: service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java [hunk 18]
```diff
@@ -100,6 +107,25 @@ public void testResourceInvalidMediaTypeResponseStatusCheck() throws Exception {
     }
 
     @Test
+     public void resourceZoneDoesNotExistException() throws Exception {
+         // NOTE: To throw a ZoneDoesNotExistException, we must ensure that the AcsRequestContext in the
+        //       SpringSecurityZoneResolver class returns a null ZoneEntity
+         MockSecurityContext.mockSecurityContext(null);
+         MockAcsRequestContext.mockAcsRequestContext(this.testZone);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java]: No need to mock request context again.

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java [hunk 19]
```diff
@@ -99,6 +106,23 @@ public void testCreatePolicyInvalidMediaTypeResponseStatusCheck() throws Excepti
                 .andExpect(status().isUnsupportedMediaType());
     }
 
+    public void policyZoneDoesNotExistException() throws Exception {
+        // NOTE: To throw a ZoneDoesNotExistException, we must ensure that the AcsRequestContext in the
+        //       SpringSecurityZoneResolver class returns a null ZoneEntity
+        MockSecurityContext.mockSecurityContext(null);
+        MockAcsRequestContext.mockAcsRequestContext(this.testZone);
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: No need to mock request context again.

File: service/src/test/java/com/ge/predix/controller/test/SubjectPrivilegeManagementControllerIT.java [hunk 20]
```diff
@@ -82,16 +85,36 @@
...
     @BeforeClass
-    public void setup() throws Exception {
+    public void setup() {
         this.testZone = TEST_UTILS.createTestZone("SubjectMgmtControllerIT");
         this.testZone2 = TEST_UTILS.createTestZone("SubjectMgmtControllerIT2");
         this.zoneService.upsertZone(this.testZone);
         this.zoneService.upsertZone(this.testZone2);
+    }
+
+    @BeforeMethod
```
Comment (irinaepshteyn) [inline service/src/test/java/com/ge/predix/controller/test/SubjectPrivilegeManagementControllerIT.java]: Refer to comments for PolicyManagementControllerIT regarding setup() and beforeMethod().  Setting up testZone2 could be deferred until the  test that is using it - testSameSubjectDifferentZones()

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
@VoollaSandeepKumar, I am done reviewing.  Please address my comments.

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java (hunk 11 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: Fixed

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java (hunk 12 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: Fixed

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java (hunk 13 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: We need to mock the request context again as this call makes changes to map that is accessed by the SpringSecurityZoneResolver based on the zone name passed to security context. The argument passed is not used inside the code so refactored the mock request context method as well

File: service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java (hunk 14 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyEvaluationControllerIT.java]: We need to mock the request context again as this call makes changes to map that is accessed by the SpringSecurityZoneResolver based on the zone name passed to security context. The argument passed is not used inside the code so refactored the mock request context method as well

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java (hunk 19 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: We need to mock the request context again as this call makes changes to map that is accessed by the SpringSecurityZoneResolver based on the zone name passed to security context. The argument passed is not used inside the code so refactored the mock request context method as well

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java (hunk 16 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: Fixed

File: service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java (hunk 15 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/PolicyManagementControllerIT.java]: Fixed

File: service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java (hunk 17 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java]: Fixed

File: service/src/test/java/com/ge/predix/controller/test/SubjectPrivilegeManagementControllerIT.java (hunk 20 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/SubjectPrivilegeManagementControllerIT.java]: Fixed

File: service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java (hunk 2 above)
Comment (VoollaSandeepKumar) [reply service/src/main/java/com/ge/predix/acs/ZoneDoesNotExistControllerAdvice.java]: Leaving the ZoneDoesNotExistControlAdvice as is will be less invasive for now.
you need to catch and throw ZonDoesNotExistException in controllers or Remove wrapping of Exceptions in controllers as RestApiExceptions

File: service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java (hunk 18 above)
Comment (VoollaSandeepKumar) [reply service/src/test/java/com/ge/predix/controller/test/ResourcePrivilegeManagementControllerIT.java]: We need to mock the request context again as this call makes changes to map that is accessed by the SpringSecurityZoneResolver based on the zone name passed to security context. The argument passed is not used inside the code so refactored the mock request context method as well

**Review by VoollaSandeepKumar — COMMENTED** [review]
Fixed your comments.

**Comment by VoollaSandeepKumar** [comment]
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1155/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1154/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1153/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/1152/

File: commons/src/main/java/com/ge/predix/acs/commons/web/GenericExceptionHandler.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice
+@Order(10)
+public class GenericExceptionHandler extends BaseRestApiControllerAdvice {
```
Comment (anubhavi25) [inline commons/src/main/java/com/ge/predix/acs/commons/web/GenericExceptionHandler.java]: Consider removing this class since Spring MVC should return an appropriate HTTP 5xx status code by default if an exception is propagated outside the controller context (i.e. not caught by a controller-level `catch` block). If a negative test that exercises this edge case doesn't already exist, consider adding it.

File: commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+
+    @ExceptionHandler(HttpMediaTypeNotAcceptableException.class)
+    @ResponseStatus(HttpStatus.NOT_ACCEPTABLE)
+    @ResponseBody
+    public RestApiErrorResponse handleException(final HttpMediaTypeNotAcceptableException e) {
+        return super.handleException(e, HttpStatus.NOT_ACCEPTABLE.getReasonPhrase());
+    }
+
+    @Override
+    public ResponseEntity handleException(final RestApiException e) {
```
Comment (anubhavi25) [inline commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java]: Pull this method definition up into the `abstract` class since it's the default implementation used by all subclasses. Also remove this method from all subclasses unless their `handleException` implementations do something different.

File: commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.HttpMediaTypeNotAcceptableException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice
+@Order(7)
```
Comment (anubhavi25) [inline commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java]: For exceptions like this one that are logically `final` (i.e. not subclassed further), I would use `@Order(Ordered.HIGHEST_PRECEDENCE)` instead of hardcoding an ordinal value as the order.

File: commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpMediaTypeNotAcceptableExceptionHandler.class);
+
+    public HttpMediaTypeNotAcceptableExceptionHandler() {
+        super(LOGGER);
+    }
+
+    @ExceptionHandler(HttpMediaTypeNotAcceptableException.class)
+    @ResponseStatus(HttpStatus.NOT_ACCEPTABLE)
+    @ResponseBody
```
Comment (anubhavi25) [inline commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java:42]: I understand the usage of the `@ResponseBody` annotation for `@RestController`-annotated classes but why is it necessary for `@ExceptionHandler`-annotated methods like this one?

**Review by anubhavi25 — CHANGES_REQUESTED** [review]
Here's a partial review of your work-in-progress changes

File: commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java (hunk 24 above)
Comment (VoollaSandeepKumar) [reply commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java:42]: @ResponseBody sets the error code and error message to response from the exception handler layer. Its needed even though we have @ResponseBody annotation at Controller level.

File: commons/src/main/java/com/ge/predix/acs/commons/web/GenericExceptionHandler.java (hunk 21 above)
Comment (VoollaSandeepKumar) [reply commons/src/main/java/com/ge/predix/acs/commons/web/GenericExceptionHandler.java]: I have tried adding a a test case. But there is not straight way to test through spring MVC. I so added a dummy endpoint and test it for possible exceptions. If there is a runtime springMVC send 500 response.Below is the sample response:

1734021-C02V3A3AHTD6:~ voolasandeepkumar$ **curl -v http://localhost:8181/v1/subject1**
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8181 (#0)
> GET /v1/subject1 HTTP/1.1
> Host: localhost:8181
> User-Agent: curl/7.54.0
> Accept: */*
> 
< **HTTP/1.1 500 Server Error**
< Content-Type: application/json;charset=utf-8
< Connection: close
< 
* Closing connection 0
{"timestamp":1511993787190,"status":500,"error":"Internal Server Error","exception":"java.lang.RuntimeException","message":"org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.RuntimeException","path":"/v1/subject1"}

File: commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java (hunk 23 above)
Comment (VoollaSandeepKumar) [reply commons/src/main/java/com/ge/predix/acs/commons/web/HttpMediaTypeNotAcceptableExceptionHandler.java]: This is fixed

**Review by anubhavi25 — APPROVED** [review]
